### ISSUE TITLE
v2.0.0-rc.2 causes 'Out of stack space' error in IE

### ISSUE THREAD

**Issue by nashwaan** [jods4/aurelia-webpack-build#33]
Hi @jods4 

The latest version of `aurelia-webpack-plugin` which is `v2.0.0-rc.2` prevents Aurelia app from running in IE.

I am using [super-minimal-aurelia](https://github.com/nathanchase/super-minimal-aurelia) setup created by @nathanchase. Installing this repo will work out of the box. But if I replace this line in `package.json`:
```json
    "aurelia-webpack-plugin": "git://github.com/aurelia/webpack-plugin.git#v2.0-rc2",
```
with this line:
```json
    "aurelia-webpack-plugin": "^2.0.0-rc.2",
```
 
And do clean installation, the app will compile but won't start in IE. I get two errors in the console:

`SCRIPT28: Out of stack space`

`SCRIPT2343: Stack overflow at line: 7803`

Any idea why?

**Comment by jods4** [comment]
No idea, that would require debugging.

I'm pretty confident that it's not caused by the Webpack plugin directly. It has very little impact on the runtime (if any) and the `v2.0-rc2` branch is exactly what was published in version `2.0.0-rc.2` so that should not make a difference.

> And do clean installation

My first guess is that another package was upgraded behind your back.

Can you post the top of the stack when it overflows? It would be interesting to see where the infinite recursion is.

**Comment by nashwaan** [comment]
> the v2.0-rc2 branch is exactly what was published in version 2.0.0-rc.2 so that should not make a difference

I have re-downloaded [super-minimal-aurelia](https://github.com/nathanchase/super-minimal-aurelia)  and before typing `npm install`, I changed `"git://github.com/aurelia/webpack-plugin.git#v2.0-rc2"` to `"^2.0.0-rc.2"`  and then did `npm install`. But then I still get same errors.

What I find it really strange is why it works perfectly when installing `aurelia-webpack-plugin` from  `"git://github.com/aurelia/webpack-plugin.git#v2.0-rc2"` and not from `"^2.0.0-rc.2"` since they are _exactly the same_!
_________
> My first guess is that another package was upgraded behind your back.

To ensure nothing is happening behind my back, I did `npm install` with `"git://github.com/aurelia/webpack-plugin.git#v2.0-rc2"` and ensured the app was working fine. Then did the below steps to ensure no other packages is doing undesired changes:

`npm uninstall --save-dev aurelia-webpack-plugin`
`npm install --save-dev aurelia-webpack-plugin@2.0.0-rc.2`

Now I don't get errors! The app works!

BUT, if I do now a clean installation (by erasing contents of `node_modules` folder and then `npm install`), I get the errors back!!!

Perhapes @nathanchase can confirm my findings.
________

> Can you post the top of the stack when it overflows? It would be interesting to see where the infinite recursion is.

As you can see in the below screenshot, the infinite recursion happens in this line:

`if (!this || this.constructor !== Collection) return new Collection(a);` 

Because this `if` statement evaluates to true and `new Collection(a)` calls iteself `function Collection(a)` and this causes infinite recursion!

![image](https://cloud.githubusercontent.com/assets/11927991/24964791/4d6d3798-1fb3-11e7-8643-41b70be8043e.png)

But.. why it works smoothly when clean-installing from github and not from npm?

**Comment by ckotzbauer** [comment]
@nashwaan Can you make a `npm shrinkwrap --dev` in both situations when the app works, and when it does not and diff the two `npm-shrinkwrap.json` files?

**Comment by jods4** [comment]
This is `aurelia-polyfills`. 
Can you check which version you have in both working and non-working setups?

**Comment by jods4** [comment]
What confuses me here is that `createCollection` has not changed for a long time.
How can the `if` evaluate to `true`?
The `new Collection(a)` call would definitely have `this && this.constructor === Collection`? What am I missing here?

**Comment by nashwaan** [comment]
@code-chris , diff of `npm-shrinkwrap.json` yeilds the following:

`npm-shrinkwrap.json` when installing `aurelia-webpack-plugin` from **github** (app working ok):
```json
    "aurelia-webpack-plugin": {
      "version": "2.0.0-rc.1",
      "from": "git://github.com/aurelia/webpack-plugin.git#v2.0-rc2",
      "resolved": "git://github.com/aurelia/webpack-plugin.git#c86042faa6dfe03ca1350662bf8e7b938523c591",
      "dev": true
    },
```

`npm-shrinkwrap.json` when installing `aurelia-webpack-plugin` from **npm** (app not working):
```json
    "aurelia-webpack-plugin": {
      "version": "2.0.0-rc.2",
      "from": "aurelia-webpack-plugin@>=2.0.0-rc.2 <3.0.0",
      "resolved": "https://registry.npmjs.org/aurelia-webpack-plugin/-/aurelia-webpack-plugin-2.0.0-rc.2.tgz",
      "dev": true
    },
```
______

@jods4 , `aurelia-polyfills` is not used here. `babel-polyfil` is used in this project. Note the versions of all other dependencies doesn't get changed in all situations.

Content of `package.json`:
```json
{
  "license": "MIT",
  "devDependencies": {
    "aurelia-webpack-plugin": "git://github.com/aurelia/webpack-plugin.git#v2.0-rc2",
    "babel-core": "^6.23.1",
    "babel-loader": "^v7.0.0-alpha.2",
    "babel-plugin-transform-decorators-legacy": "^1.3.4",
    "babel-polyfill": "6.23.0",
    "babel-preset-env": "^1.3.2",
    "babel-preset-stage-0": "^6.22.0",
    "babili-webpack-plugin": "^0.0.11",
    "core-js": "^2.4.1",
    "css-loader": "^0.28.0",
    "html-loader": "^0.4.5",
    "html-webpack-plugin": "^2.28.0",
    "mutationobserver-shim": "^0.3.2",
    "raf": "^3.3.0",
    "style-loader": "^0.16.1",
    "webpack": "^2.2.1",
    "webpack-dev-server": "^2.4.1"
  },
  "dependencies": {
    "aurelia-bootstrapper": "^2.1.1"
  },
  "scripts": {
    "web": "webpack-dev-server --hot --inline",
    "prod": "webpack -p --progress --profile --colors"
  }
}
```

**Comment by jods4** [comment]
`aurelia-polyfills` is always used, it's loaded by `aurelia-bootstrapper`.

**Comment by nashwaan** [comment]
@jods4 

some content of `node_modules\aurelia-polyfills\package.json` (app working ok):
```json
  "_requested": {
    "raw": "aurelia-polyfills@^1.0.0",
    "scope": null,
    "escapedName": "aurelia-polyfills",
    "name": "aurelia-polyfills",
    "rawSpec": "^1.0.0",
    "spec": ">=1.0.0 <2.0.0",
    "type": "range"
  },
  "_requiredBy": [
    "/aurelia-bootstrapper"
  ],
  "_resolved": "https://registry.npmjs.org/aurelia-polyfills/-/aurelia-polyfills-1.2.1.tgz",
  "_shasum": "c429385691f39cb0bf54862385995532d9725411",
  "_shrinkwrap": null,
  "_spec": "aurelia-polyfills@^1.0.0",
```

some content of `node_modules\aurelia-polyfills\package.json` (app not working):
```json
  "_requested": {
    "raw": "aurelia-polyfills@https://registry.npmjs.org/aurelia-polyfills/-/aurelia-polyfills-1.2.1.tgz",
    "scope": null,
    "escapedName": "aurelia-polyfills",
    "name": "aurelia-polyfills",
    "rawSpec": "https://registry.npmjs.org/aurelia-polyfills/-/aurelia-polyfills-1.2.1.tgz",
    "spec": "https://registry.npmjs.org/aurelia-polyfills/-/aurelia-polyfills-1.2.1.tgz",
    "type": "remote"
  },
  "_requiredBy": [
    "/",
    "/aurelia-bootstrapper"
  ],
  "_resolved": "https://registry.npmjs.org/aurelia-polyfills/-/aurelia-polyfills-1.2.1.tgz",
  "_shasum": "c429385691f39cb0bf54862385995532d9725411",
  "_shrinkwrap": null,
  "_spec": "aurelia-polyfills@https://registry.npmjs.org/aurelia-polyfills/-/aurelia-polyfills-1.2.1.tgz",
```

**Comment by jods4** [comment]
Really not sure what happens here.

Maybe it would help to better understand the cause of the problem.
As I wrote above, I don't see how `new X()` can possibly satisfy `if (!this || this.constructor !== X)`.

The other track to investigate is, based on your answer:
> To ensure nothing is happening behind my back, I did npm install with "git://github.com/aurelia/webpack-plugin.git#v2.0-rc2" and ensured the app was working fine. Then did the below steps to ensure no other packages is doing undesired changes:
npm uninstall --save-dev aurelia-webpack-plugin
npm install --save-dev aurelia-webpack-plugin@2.0.0-rc.2
Now I don't get errors! The app works!
BUT, if I do now a clean installation (by erasing contents of node_modules folder and then npm install), I get the errors back!!!

To me, that is proof that another lib gets updated behind your back and causes this issue. It would be nice to identify which one.

**Comment by nashwaan** [comment]
After various attempts and experimentation, it turned out that `babel-polyfill` was causing all the issues here (although I still don't understand why it worked when installing aurelia-webpack-plugin from github but not from npm). I just removed it completely from [super-minimal-aurelia](https://github.com/nathanchase/super-minimal-aurelia) and everything works fine now:

```js
module.exports = {
  entry: {
    main: [ 
      './ie-polyfill',
      // 'babel-polyfill', // the problematic and redundant polyfill
      'aurelia-bootstrapper'
    ]
  },
```

I have submitted a pull request in that repo to fix this issue so it can work directly from `"aurelia-webpack-plugin": "^2.0.0-rc.2"`.

Thanks @jods4 and @code-chris for your efforts.
